### ISSUE TITLE
[Bug]: Filtering results do not match expectations when using the radius param.

### ISSUE THREAD

**Issue by syjzlee** [milvus-io/milvus#30327]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: v2.3.3
- Deployment mode(standalone or cluster): standalone
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): Ubuntu
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. I can get the second record whose id equal "71a5b70e-bcc1-11ee-ae22-708bcdba5a7a" when I set nothing to radius param.
![1](https://github.com/milvus-io/milvus/assets/21698608/04a21e78-2538-47fe-bc0f-62c681c132d0)

2. But I can't see it anymore when i set 0.7 to radius. Although I have no delete the that record.

![2](https://github.com/milvus-io/milvus/assets/21698608/560e7745-a7d9-4a49-8ed4-5eb6a205fed1)

My index is IP. Maybe I'm wrong about the radius parameter? If my understanding is wrong, can you tell me how to filter the result with score greater than 0.7?
![3](https://github.com/milvus-io/milvus/assets/21698608/346d952d-a4cf-4297-86ab-d1a2a87067d2)

### Expected Behavior

Input some param to filter the result with score greater than 0.7.

**Comment by smellthemoon** [comment]
If it's convenient, could you provide more detailed logs for further investigation？
@syjzlee

**Comment by syjzlee** [comment]
There are no server logs. Only parameters used by the visual client are available through the Attu.

**Comment by Asperas13** [comment]
I can confirm i have a similar behaviour with “Hamming” metric
When i manually filter results like this:
result = []
for hit in hits:
    If hit.distance < radius:
          result.append(hit)
I’m able to see all similar entities, but this works slower
Hovewer when i use “radius” search param then some closes neighbours disapear

**Comment by mkhaliki24** [comment]
I seem to have a problem related to this.

- Milvus version: v2.3.4
- Deployment mode(standalone or cluster):
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2): milvus-sdk-java 2.3.4
- OS(Ubuntu or CentOS):
- CPU/Memory: 
- GPU: 
- Others:

index_params = {
        "index_type": "IVF_SQ8",
        "params": {"nlist": 1024},
        "metric_type": "COSINE"
}

### Current Behavior

1. Search with COSINE metric_type and params:

![Screenshot 2024-03-14 at 2 03 17 PM](https://github.com/milvus-io/milvus/assets/163457458/1b8831dd-7bed-4670-9d23-e43d26d666b8)

Results:
![Screenshot 2024-03-14 at 2 02 53 PM](https://github.com/milvus-io/milvus/assets/163457458/b41936ab-2213-46a2-9ae8-7cdf9b2dbfe7)

2. Same search but with higher radius (0.7)

![Screenshot 2024-03-14 at 2 08 31 PM](https://github.com/milvus-io/milvus/assets/163457458/14c89557-93e2-47a8-9824-bff00e24afd1)

Results:
![Screenshot 2024-03-14 at 2 08 41 PM](https://github.com/milvus-io/milvus/assets/163457458/4d75f978-ddbd-4c47-a4f7-bcff6b94b8f8)

### Expected Behavior
In the first search result, the first record we get back has a distance of 0.80049676. Increasing the radius from 0.1 to 0.7 should still return the first record.

### Steps To Reproduce
No response

### Milvus Log
No response

### Anything else?
No response

**Comment by maggonravi** [comment]
Is there a resolution for this? Facing exact same issue when using radius with IP.

**Comment by yanliang567** [comment]
could you please file a new issue and offer the milvus logs for investigation?

**Comment by xiaofan-luan** [comment]
@smellthemoon  @cydrain 
is there a knob we can increase nprobe and improve the searc quality?

**Comment by xiaofan-luan** [comment]
I don't think nprobe will work on this case

**Comment by cydrain** [comment]
Hi @mkhaliki24 @maggonravi,

For IVF-serial index types, you can try this param "max_empty_result_buckets" to fine tuning their range search results.
![Screenshot from 2024-06-19 11-11-11](https://github.com/milvus-io/milvus/assets/3992404/3b465752-c239-4159-8fc1-8a51d7d32850)

**Comment by maggonravi** [comment]
Hi @cydrain ,

I tried setting max_empty_result_buckets to low value (10) and to max value (65535) but still got empty results.

Here are the params. If I remove range_filter and radius, it returns 5 documents from db. If I pass those, I get empty. While metric has value of ~ .7 - .8. I tried varying radius between -1 to 1 but no help.

```
{
    'nprobe': 128, 
    'range_filter': 1.0, 
    'radius': -1, 
    'max_empty_result_buckets': 65535
}
```

**Comment by yanliang567** [comment]
@maggonravi could you please file a new issue and offer the milvus logs? if possible sharing some data sample with us to reproduce the issue would be great helpful.

**Comment by xiaofan-luan** [comment]
> Hi @cydrain ,
> 
> I tried setting max_empty_result_buckets to low value (10) and to max value (65535) but still got empty results.
> 
> Here are the params. If I remove range_filter and radius, it returns 5 documents from db. If I pass those, I get empty. While metric has value of ~ .7 - .8. I tried varying radius between -1 to 1 but no help.
> 
> ```
> {
>     'nprobe': 128, 
>     'range_filter': 1.0, 
>     'radius': -1, 
>     'max_empty_result_buckets': 65535
> }
> ```

what metrics you are using? 
For cosine metrics, 1.0 means most similar and you just fitered out all the vectors

**Comment by mprudra** [comment]
> For cosine metrics, 1.0 means most similar and you just fitered out all the vectors

As per the official docs, the above filter seems correct to me.
`radius` is the actual threshold of the metric and `range_filter` is the secondary filter to even exclude closest ones, that's what I could gather from the docs.

![image](https://github.com/milvus-io/milvus/assets/161220745/e1266416-5294-429a-8ef7-742eee61d11a)

So  `radius` (-1) < distance <= `range_filter` (1.0)
This should have resulted in some results.

**Comment by cydrain** [comment]
Hi @mprudra ,

**Comment by cydrain** [comment]
Hi @mprudra ,

I can reproduce this issue in my local machine.
I have filed another issue to track https://github.com/milvus-io/milvus/issues/34199

**Comment by cydrain** [comment]
Hi @mprudra ,
This issue has been fixed in Milvus v2.4.x release, you can see my comments in #34199

**Comment by subhamhimself** [comment]
Hi @cydrain We're re-experiencing this issue.
I'm not sure, but it could be because of Milvus upgrade to 2.5

This is happening for Collection.search_iterator.
We're setting: `{"radius": -1, "range_filter": 1}`, and I tried a variety of other values but no use.

Error I get:
```
Cannot init search iterator because init page contains no matched rows, please check the radius and range_filter set up by searchParams
```
The partition has well over a million vectors.

Can you please check if you can reproduce this issue? 
cc @maggonravi
